planner/core/tiflash_selection_late_materialization.go

```diff
@@ -0,0 +1,257 @@
+// and sort the groups by the selectivity of the conditions in the group.
+// @param: conds: the conditions to be grouped
+// @return: the groups of conditions sorted by the selectivity of the conditions in the group
+// @example: conds = [a > 1, b > 1, a > 2, c > 1, a > 3, b > 2], return = [[a > 3, a > 2, a > 1], [b > 2, b > 1], [c > 1]]
+// @note: when the selectivity of one group is larger than the threshold, we will remove it from the returned result.
+// @note: when the number of columns of one group is larger than the threshold, we will remove it from the returned result.
+func groupByColumnsSortBySelectivity(sctx sessionctx.Context, conds []expression.Expression, physicalTableScan *PhysicalTableScan) []expressionGroup {
+	// Create a map to store the groupMap of conditions keyed by the columns
+	groupMap := make(map[string][]expression.Expression)
+
+	// Iterate through the conditions and group them by columns
+	for _, cond := range conds {
+		// If excuting light cost condition first,
+		// the rows needed to be exucuted by the heavy cost condition will be reduced.
+		// So if the cond contains heavy cost functions, skip it.
+		if withHeavyCostFunction(cond) {
+			continue
+		}
+
+		columns := expression.ExtractColumns(cond)
+
+		var code string
+		if len(columns) == 0 {
+			code = "0"
+		} else if len(columns) <= columnCountThreshold {
+			code = transformColumnsToCode(columns, len(physicalTableScan.Columns))
+		} else {
+			// If the number of columns is larger than columnCountThreshold,
+			// the possibility of the selectivity of the condition is larger than the threshold is very small.
+			// So we skip it to reduce the cost of calculating the selectivity.
+			continue
+		}
+		groupMap[code] = append(groupMap[code], cond)
+	}
+
+	// Estimate the selectivity of each group and check if it is larger than the selectivityThreshold
+	var exprGroups []expressionGroup
+	for _, group := range groupMap {
+		selectivity, _, err := physicalTableScan.tblColHists.Selectivity(sctx, group, nil)
+		if err != nil {
+			logutil.BgLogger().Warn("calculate selectivity failed, do not push down the conditions group", zap.Error(err))
+			continue
+		}
+		if selectivity <= selectivityThreshold {
+			exprGroups = append(exprGroups, expressionGroup{exprs: group, selectivity: selectivity})
+		}
+	}
+
+	// Sort exprGroups by selectivity in ascending order
+	sort.SliceStable(exprGroups, func(i, j int) bool {
+		return exprGroups[i].selectivity < exprGroups[j].selectivity || (exprGroups[i].selectivity == exprGroups[j].selectivity && len(exprGroups[i].exprs) < len(exprGroups[j].exprs))
+	})
+
+	return exprGroups
+}
+
+// withHeavyCostFunction is used to check if the condition contain heavy cost functions.
+// @param: cond: condition of PhysicalSelection to be checked
+// @note: heavy cost functions are functions that may cause a lot of memory allocation or disk IO.
+func withHeavyCostFunction(cond expression.Expression) bool {
```

Comment:
```suggestion
func withHeavyCostFunctionForTiFlashPrefetch(cond expression.Expression) bool {
```

I prefer to make it more straightforward. The `heavy` here cannot be used in other places in the planner, like that decide whether a filter should be pushed down or not by whether it's heavy or not. It's a particular case of late materialization.

For example, `CASE WHEN` should be considered heavy, but `DateSub` will not be regarded as a heavy filter. 

And just a curious question, if the heavy cost function you mentioned here has an excellent selectivity, we still do not consider it a possible choice of late materialization?